File: packages/shared/src/crypto/canonicalJson.ts
/** Deterministic JSON serialization: sorts keys recursively. */
export function canonicalJson(obj: unknown): string {
  return JSON.stringify(sortKeys(obj));
}

function sortKeys(val: unknown): unknown {
  if (val === null || val === undefined) return val;
  if (Array.isArray(val)) return val.map(sortKeys);
  if (typeof val === "object") {
    const sorted: Record<string, unknown> = {};
    for (const k of Object.keys(val as Record<string, unknown>).sort()) {
      sorted[k] = sortKeys((val as Record<string, unknown>)[k]);
    }
    return sorted;
  }
  return val;
}

File: apps/api/app/canonical_json.py
"""Deterministic JSON serialization: sorts keys recursively.

Must produce byte-identical output to the TypeScript version
(packages/shared/src/crypto/canonicalJson.ts).
"""

from __future__ import annotations

import json
from typing import Any


def _sort_keys(val: Any) -> Any:
    if val is None:
        return val
    if isinstance(val, list):
        return [_sort_keys(v) for v in val]
    if isinstance(val, dict):
        return {k: _sort_keys(v) for k, v in sorted(val.items())}
    return val


def canonical_json(obj: Any) -> str:
    """Return deterministic JSON string with sorted keys, no extra whitespace."""
    return json.dumps(_sort_keys(obj), separators=(",", ":"), ensure_ascii=False)
